Fix TabBar Hidden Tabs Handling

lodetrick opened the pull request:
See: #106164

These bugfixes are related but necessary for the above PR to function.

The first fix is in TabContainer. Currently TabContainer calls `_update_margins` in the `add_child_notify` method. Since it adds the `TabBar` as a child upon being allocated and the theme cache is not set when allocating the `EditorBottomPanel` class, this leads to a segfault as it is trying to access a nullptr. This PR fixes the bug by wrapping the theme access behind an if statement, removing many redundant ternaries.

The second fix is in TabBar's theme constants. The `tab_separation` theme variable did not work correctly where hidden tabs intermixed with non-hidden tabs could still add their separation even though they are not visible. This PR solves the issue by only adding the separation for non-hidden tabs.

lodetrick committed:
Fix TabBar hidden tabs handling

kitbdev reviewed (commented):
The tab_separation changes also need to be done in the draw selected tab (below the draw unselected tab) and in `TabBar::get_tab_rect`. 

It looks like the segfault only happened in a previous commit of #106164, so it is hard to debug. There is probably a deeper issue there that caused it. The current changes to TabContainer look fine.

lodetrick commented:
Ok, I took a deeper look at how the `tab_separation` worked and realized that we should just always use the `ofs_cache` when drawing the tabs, we don't need to recalculate it and add more redundant code. I realized that a lot of the if statements were bandaid fixes that could be removed once the ofs_cache was correct. This ended up cleaning the code related to the `tab_separation` a lot, removing all but five instances of the variable in the code while still acting the same.

lodetrick commented:
Alright, what I did:

I reworked how it calculates where to place the dropping texture (the bar that shows up between tabs).  Instead of calculating the average distance between tabs one apart, it merely adds and subtracts half the `tab_separation`, which is defined to be the distance between tabs. This removed a rather complicated bundle of if-statements and made the code structure closer to how it was before `tab_separation` was implemented.

I took the opportunity to add the `get_previous_available` and `get_next_available` public methods (making the `select_previous_available` use these internally), these methods are really useful in cases where there are hidden tabs interleaved. Bringing it back to #106164, these methods are necessary for the functionality to move tabs using buttons through the dock manager popup, as the bottom dock is structured to have interleaved hidden tabs. There are no docs for these yet, if this passes checks and is not urgent then I'll let a future PR decide if we want to expose these methods.

I also reworked the `get_closest_tab_idx_to_point` method. Previously it used the Vector2.distance function but since the TabBar only works on an axis we can just compute `abs(center - p_point.x)`. This is an easy performance increase, and if vertical tabs are implemented we only need to change the axis.

lodetrick commented on scene/gui/tab_bar.cpp:
```diff
@@ -35,6 +35,7 @@
 #include "scene/gui/texture_rect.h"
 #include "scene/main/viewport.h"
 #include "scene/theme/theme_db.h"
+#include <cstdlib>
```
Shoot, didn't mean to include that

kitbdev commented on scene/gui/tab_bar.cpp:
```diff
@@ -1543,28 +1531,35 @@ void TabBar::_move_tab_from(TabBar *p_from_tabbar, int p_from_index, int p_to_in
-		for (int i = offset; i <= max_drawn_tab; i++) {
-			Vector2 center = get_tab_rect(i).get_center();
-			float distance = center.distance_to(p_point);
+	int closest_tab = -1;
+	float closest_distance = FLT_MAX;
+
+	// Search along the x-axis since the TabBar is horizontal.
+	for (int i = offset; i <= max_drawn_tab; i++) {
+		if (!tabs[i].hidden) {
+			float center = get_tab_rect(i).get_center().x;
+			float distance = Math::abs(center - p_point.x);
```
I think this changes the behavior when over non-equal sized tabs, since you removed the `get_tab_idx_at_point` check.

lodetrick replied:
Thats a good point. I think before it would have prioritized the early-indexed tabs, so I'll re-add the `get_tab_idx_at_point` to account for that.

kitbdev commented on scene/gui/tab_bar.cpp:
```diff
@@ -588,38 +575,20 @@ void TabBar::_notification(int p_what) {
-							x = (tab_rect.position.x + prev_tab_rect.position.x + prev_tab_rect.size.width) / 2;
-						} else { // First tab, will appear on left edge.
-							x = tab_rect.position.x;
+					if (get_local_mouse_position().x > tab_rect.get_center().x) {
```
There's an issue with the hover preview, I think since it now gets the hovered tab first, not always the closest one.

https://github.com/user-attachments/assets/c5095d7f-239d-41f0-8b00-94f18f2ab3eb

lodetrick replied:
I'm haivng some trouble recreating this specific issue (I did notice a similar one where the hover still showed a phantom tab off the end of the tabbar, looking into that). What settings/files are you using?

lodetrick replied:
Found it. Having `get_closest_tab_idx_to_point` call `get_tab_idx_at_point` caused the issue bc `get_tab_idx_at_point` still looked at the rects of hidden tabs, which should not be happening. I modified `get_tab_idx_at_point` to only check tabs that are not hidden, and that fixed the issue.

kitbdev commented on scene/gui/tab_bar.cpp:
```diff
@@ -1212,6 +1196,9 @@ void TabBar::_update_cache(bool p_update_hover) {
+			w += theme_cache.tab_separation;
 		}
 
 		tabs.write[i].accessibility_item_dirty = true;
```
This isn't an issue with your changes, but this line should be before all the continues. It should run on all tabs, since it needs to be triggered on a tab when it is hidden or just moved, judging by the code in `NOTIFICATION_ACCESSIBILITY_UPDATE`.
If you don't want to change it in this PR, I can open a separate PR for it.

lodetrick replied:
I can do it in this PR, I need to rename the commit anywho.

kitbdev replied:
I don't think you can test it without a screen reader. Looks good, I tested it and the behavior is much better than before.

kitbdev reviewed (approved):
Looks good.

I think we should update the PR and commit title to be more accurate, since it does more than change the drawing logic. The segfault is more of a potential segfault too since it doesn't really happen in master. Something like "Fix TabBar hidden tabs handling".

Repiteo commented:
Thanks!
